Migrate Mirage 2 build from bower and Ruby sass to npm and node-sass

alanorth opened the pull request:
## References
* [Recent mailing list post about Mirage 2 build issues related to torquebox.org](https://groups.google.com/g/dspace-tech/c/7_fUBxsxkhs/m/xTpRzyDUAQAJ)

## Description
In short, this pull request replaces bower and Ruby sass with npm and node-sass in order to fix the broken Mirage 2 build when using the "slow" maven build. This replaces the [previous pull request that was using yarn](https://github.com/DSpace/DSpace/pull/8288).

For more background, the Mirage 2 build currently uses bower for package management and Ruby sass/compass for Sass processing. Bower has been deprecated for years and is an increasing source of obscure build issues. Compass is Ruby based and requires the user to either preinstall Ruby and several gems or use those provided by JRuby via torquebox.org during build time. As of last week the torquebox.org domain expired and is likely abandoned, making it impossible to build Mirage 2.

## Instructions for Reviewers
Test building Mirage 2 at your site using the `-Dmirage2.on=true -Dmirage2.deps.included=true` option to maven. I have tested both vanilla Mirage 2 and a slightly modified one with style overrides in `dspace/modules/xmlui-mirage2`.

## Notes
List of changes in this PR:
*  Allow building Mirage 2 with Node.js 12 and use eirslett/frontend-maven-plugin to install Node.js 12 during maven "slow" build mode. JavaScript packages that were previously installed by bower are now installed by npm.
* Update CI to use latest Node.js 12 and remove useless pre-requisite step (as GitHub CI already has npm and grunt installed).
* Remove JRuby and compass stuff from build and use node-sass as the Sass processor and essentially polyfill the compass Sass mixins using compass-mixins from npmjs.com.
* Use a static copy of Modernizr.min.js because the version we need is not available on npmjs.com.
* Update grunt build tools

## TODO
Once this is merged we will also need to update the DSpace installation documentation on the wiki.

alanorth committed:
dspace/modules/xmlui-mirage2/pom.xml: update npm tasks

Update frontend-maven-plugin and use the default npm provided by
the distribution rather than specifying a version explicitly.

Also, add a minor tweak to the copy-resources goal so that our
overrides in dspace/modules/xmlui-mirage2 don't get over written
by the default Mirage2 source.

alanorth committed:
dspace-xmlui-mirage2: build Mirage 2 without Ruby

Update Mirage 2 so it can build without the Ruby-based sass and
compass:

- Port bower.json dependencies to package.json
- Polyfill compass's sass mixins in _main.scss from compass-mixins
- Adjust the Gruntfile.js to use node-sass* instead of Ruby sass and
  add task for copying Bootstrap web fonts
- Allow building on newer Node.js (LTS 12 for now)

Note: Node.js 12.x does not seem to support node-sass 8.x so we use
7.x here.

alanorth committed:
dspace-xmlui-mirage2: use a static copy of modernizer.min.js

Long story short: this was previous installed by bower, but bower
is deprecated and we are trying to move dependencies to npm. The
version here is 2.8.3, which is six years old and does not exist
on npm's registry, and can't be installed from git because it is
missing package.json. Since the file is not changing whatsoever
I think it's fine if we just keep a static copy in the tree and
bundle it in with the theme.js using grunt, which has the added
benefit of doing one less network request.

alanorth committed:
dspace-xmlui-mirage2: fix path to IE 9 scripts

These are now installed by npm and we place them in the scripts
directory with grunt.

alanorth committed:
dspace-xmlui-mirage2: fix path to JavaScript assets

The modules installed via npm are organized slightly differently
than they were when installed via bower.

alanorth committed:
dspace/modules/xmlui-mirage2: remove rubygems

The rubygems.torquebox.org domain is no longer active so it is not
possible to retrieve these gems during build (and we're not using
them anyways, since we switched to node-sass).

alanorth committed:
xmlui: Remove remaining Ruby and compass stuff

We no longer need the Ruby-based sass and compass to build Mirage2.

alanorth committed:
xmlui: Update slow Mirage2 build

Make sure we run the correct grunt tasks when running the "slow"
Mirage2 build with `-Dmirage2.deps.included=true`.

alanorth committed:
dspace-xmlui-mirage2: Fix glyphicon encoding

This is due to using node-sass instead of Ruby compass. The header
trail shows Â instead of /.

See: https://github.com/twbs/bootstrap-sass/issues/919

alanorth committed:
xmlui: remove sass.version var from pom.xml

alanorth committed:
dspace-xmlui-mirage2/pom.xml: exclude modernizr.min.js

Exclude our static modernizr.min.js from license check.

alanorth committed:
.github: update Mirage2 build

We no longer use Ruby-based SASS processors.

alanorth committed:
.github/workflows/build.yml: Use setup-node

Use setup-node action to install Node.js instead of a bash command
with nvm. For some reason the old nvm method keeps giving me Node
16...

See: https://github.com/actions/setup-node

alanorth committed:
dspace-xmlui-mirage2: update jQuery to 3.4.1

We bumped this in bower.json earlier in 6.4-SNAPSHOT development.

See: https://github.com/DSpace/DSpace/pull/2918

alanorth committed:
dspace-xmlui-mirage2: remove bower.json

We no longer use bower so this file is not needed.

alanorth committed:
Use latest Node.js 12.x

For CI we can simply use "12" which will take the latest 12.x LTS
release rather than pinning 12.22.4. For the normal maven build
we need to specify an actual version number because that is how
the frontend-maven-plugin works.

Also update comment about keeping these versions in sync in pom.xml
and .github/workflows/build.yml, instead of travis.yml.

alanorth committed:
dspace/modules/xmlui-mirage2/pom.xml: remove bower comment

alanorth committed:
.github/workflows/build.yml: remove pre-requisites task

We will get npm from the Node.js 12.x distribution, and GitHub CI
already has grunt installed, so there's actually nothing else we
need to install in order to build Mirage 2.

alanorth committed:
dspace-xmlui-mirage2: update readme.md

Remove all references to Ruby and update install instructions to
mention npm instead of bower. Also clean up some formatting and
plaintext http links as well as some minor style edits.

alanorth committed:
dspace-xmlui-mirage2: update grunt deps in package.json

This bumps the version of several grunt tools to their latest, with
the exception of grunt-contrib-handlebars because our handlebars is
old.

mwoodiupui commented on dspace-xmlui-mirage2/pom.xml:
```diff
@@ -53,6 +53,7 @@
                         <exclude>**/readme.txt</exclude>
                         <exclude>**/.bowerrc</exclude>
```
Do we still need `.bowerrc`?  Maybe remove here and in `dspace-xmlui-mirage2/src/main/webapp`?

alanorth replied:
Good catch! I've removed them.

kshepherd commented on dspace/modules/xmlui-mirage2/src/main/webapp/themes/Mirage2/styles/_style.scss:
```diff
@@ -58,4 +58,8 @@
+}
+
+body {
+    color: green;
```
This file looks like an unintended inclusion from a local theme?

alanorth replied:
Good catch, yes that was added accidentally with `git add -u`. Removed!

kshepherd reviewed (commented):
Looks great to me, thakns @alanorth  
I've tested a build locally and it builds quickly and successfully, and Mirage 2 is working
I left a comment about a "color:green" added to body text in the Mirage 2 _style.scss overlay which I wasn't quite sure about.

Once that little bit is cleared up I'm good to approve!

alanorth committed:
Remove unneeded .bowerrc

kshepherd reviewed (approved):
+1 reviewed and tested

kshepherd commented:
I was just trying this out with a heavily-customised repo I help with and it made me think we should include some notes in the documentation or release notes about how some of these changes work, so people with custom themes that might be importing extra node deps, including other scss with references to the old paths to deps, etc, can update their local themes to match the new way dependencies are managed

alanorth committed:
dspace-xmlui-mirage2: update readme.md

Add some notes about migrating dependencies from bower.json to the
new npm package.json.

Remove references to `compass watch` since that doesn't work anymore.

johnpinto1 commented:
A patch for the 21 commits fixed the build for Edinburgh Datashare. Thanks @alanorth. I hope it will be merged into the Dspace 6 branch Dspace: dspace_6.x soon.

alanorth commented:
Thank you, @johnpinto1. I suspect you had to move some dependencies over from `.bowerrc` to `package.json`, update some include paths from `vendor` to `node_modules`, and some include paths in SCSS? I tried to capture some of those tips in the [updated Mirage 2 readme.md](https://github.com/DSpace/DSpace/pull/8292/commits/7bc99ec88a4240349969272de35b81dd207148e6) included in this PR.

I think we are close to merging this now, thanks to your +1. Also, I have tested locally with Node.js 14 and it works, so I will push to get that in here too since Node.js 12 is no longer supported upstream.

tdonohue reviewed (approved):
👍 The code here all looks reasonable to me.  I haven't been able to test it myself, but I trust the tests of others (and it sounds like at least 3 others have tested it successfully)

tdonohue commented:
NOTE: This issue was also fixed on `dspace-5_x`, but in a different manner. In that branch, I simply found a different way to pull in the required RubyGems, see this commit which provided the fix: https://github.com/DSpace/DSpace/commit/0428da8dab7a02c592ec96ab9cea545f5b6de42d
